### ISSUE TITLE
AMICI installation on windows

### ISSUE THREAD

**Issue by arnabmutsuddy** [AMICI-dev/AMICI#1505]
I've been trying to install amici on a windows machine using Anaconda. I followed the instructions as described [here](https://amici.readthedocs.io/en/latest/python_installation.html) to install MinGW, OpenBLAS and SWIG, and added the suggested environment variables. Then I created a new conda environment and used the following command to install amici:

pip install --global-option="build_clib" --global-option="--compiler=mingw32" --global-option="build_ext" --global-option="--compiler=ming32" amici --no-cache-dir --verbose

But installation fails with a rather elaborate error message. Is there anything I should have done differently?

[amici_error_log1.txt](https://github.com/AMICI-dev/AMICI/files/6446603/amici_error_log1.txt)

**Comment by thomassligon** [comment]
I see mingw32 and ming32. The error log has code searching for msvc, which is default, so ming32 might be a problem.

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________
From: arnabmutsuddy ***@***.***>
Sent: Saturday, May 8, 2021 11:27:09 PM
To: AMICI-dev/AMICI ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [AMICI-dev/AMICI] AMICI installation on windows (#1505)


I've been trying to install amici on a windows machine using Anaconda. I followed the instructions as described here<https://amici.readthedocs.io/en/latest/python_installation.html> to install MinGW, OpenBLAS and SWIG, and added the suggested environment variables. Then I created a new conda environment and used the following command to install amici:

pip install --global-option="build_clib" --global-option="--compiler=mingw32" --global-option="build_ext" --global-option="--compiler=ming32" amici --no-cache-dir --verbose

But installation fails with a rather elaborate error message. Is there anything I should have done differently?

amici_error_log1.txt<https://github.com/AMICI-dev/AMICI/files/6446603/amici_error_log1.txt>

**Comment by arnabmutsuddy** [comment]
Update: tried again without mingw32, now I'm seeing a different error. It says it can't find something called 'cblas.h'

[amici_error_log2.txt](https://github.com/AMICI-dev/AMICI/files/6448898/amici_error_log2.txt)

**Comment by thomassligon** [comment]
I can see the following output at the end of the error log:
`  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Iamici/include -Iamici/ThirdParty/SuiteSparse/KLU/Include/ -Iamici/ThirdParty/SuiteSparse/AMD/Include/ -Iamici/ThirdParty/SuiteSparse/COLAMD/Include/ -Iamici/ThirdParty/SuiteSparse/BTF/Include/ -Iamici/ThirdParty/SuiteSparse/SuiteSparse_config -Iamici/ThirdParty/SuiteSparse/include -Iamici/ThirdParty/sundials/include -Iamici/ThirdParty/sundials/src -Iamici/ThirdParty/gsl/ -IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\include -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt /EHsc /Tpamici\src\cblas.cpp /Fobuild\temp.win-amd64-3.9\amici\src\cblas.obj /IC:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10 /std:c++14
  cblas.cpp
  amici\src\cblas.cpp(16): fatal error C1083: Cannot open include file: 'cblas.h': No such file or directory
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.28.29910\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2`

The entry
`/IC:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10`
shows me that you have created the environment variable that tells the compiler where to find cblas.h, but it is not finding it. It looks like something is missing in your installation of openBLAS. Does the directory `C:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10` contain a copy of cblas.h?

The entry
` C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\bin\HostX86\x64\cl.exe`
shows me that the python distutils are using MSVC (Microsoft Visual Studio) to compile and link the amici module cblas.cpp (which requires cblas.h), but I thought you were using mingw32.

**Comment by thomassligon** [comment]
Yes, I would say that installation of openBLAS did not succeed. When you run installOpenBLAS.ps1, which calls compileBLAS.cmd, you need to make sure that the sources of openBLAS are compiled and linked correctly, and the result ends up in the directory that your environment variable is pointing to. I also noticed that line 3 of the script compileBLAS.cmd calls the MSVC compiler, so you should be aware of that. In our previous exchange, we saw that the MSVC compiler was being used for amici. In fact, that is the compiler that I used (I have never used mingw32), but I think there is a big danger in mixing two different C compilers, because they have completely different runtime libraries, so you should also make sure that you use the same compiler for all tasks.

**Comment by arnabmutsuddy** [comment]
I did use mingw32 initially but decided to discard it as it didn't work for me. But the files for mingw32 are still there, could those affect my future installations of AMICI even if I decide to use MSVC only?

As per my last attempt at installing OpenBLAS, I took notice of line 3 of compileBLAS.cmd and edited it to include my current version of MSVC (2019). However, installOpenBLAS.ps1 seems to fail at line 13. From what I understand, it copies the file openblas.lib from \lib\release within the OpenBLAS directory. I've looked in the OpenBLAS folder that was extracted by the powershell script and saw that there's no folder called 'lib' within that directory. In fact, I couldn't find any file called 'openblas.lib' there. Does that indicate complieBLAS.cmd wasn't called correctly?

**Comment by arnabmutsuddy** [comment]
If I've built OpenBLAS using the installOpenBLAS powershell script, do I still need to install BLAS locally in the conda environment?

**Comment by FFroehlich** [comment]
> If I've built OpenBLAS using the installOpenBLAS powershell script, do I still need to install BLAS locally in the conda environment?

As long as you set the respective environment variables, you shouldn't need to install anything with conda. In fact, we had the experience that installing something in conda can easily interfere with other installations.

**Comment by arnabmutsuddy** [comment]
Alright, I've removed all traces of Anaconda from my system, built BLAS and added the environment variables (BLAS_CFLAGS and BLAS_LIBS), installed SWIG and added to path but somehow still end up with the same error when trying to install AMICI:

  amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_dgemv
  amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_dgemm
  amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_daxpy
  build\lib.win-amd64-3.8\amici\_amici.cp38-win_amd64.pyd : fatal error LNK1120: 3 unresolved externals
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.28.29910\\bin\\HostX86\\x64\\link.exe' failed with exit status 1120

Is there any other environment variables that I might be missing?

[error_log_full.txt](https://github.com/AMICI-dev/AMICI/files/6467987/error_log_full.txt)

**Comment by FFroehlich** [comment]
> Alright, I've removed all traces of Anaconda from my system, built BLAS and added the environment variables (BLAS_CFLAGS and BLAS_LIBS), installed SWIG and added to path but somehow still end up with the same error when trying to install AMICI:
> 
> amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_dgemv
> amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_dgemm
> amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_daxpy
> build\lib.win-amd64-3.8\amici_amici.cp38-win_amd64.pyd : fatal error LNK1120: 3 unresolved externals
> error: command 'C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\bin\HostX86\x64\link.exe' failed with exit status 1120
> 
> Is there any other environment variables that I might be missing?
> 
> [error_log_full.txt](https://github.com/AMICI-dev/AMICI/files/6467987/error_log_full.txt)

No there aren't any additional environment variables that need to be set. Comparing to https://github.com/AMICI-dev/AMICI/runs/2450686166?check_suite_focus=true , the linker invocation looks correct. Two possible sources of error that come to mind:

- openBLAS and AMICI need to be compiled with exactly the same compiler, are you sure this is the case?
- can you confirm there is a file `openblas.lib` located in `C:\BLAS\lib`?

If both are true then we are probably left to inspecting the symbol tables themselves to figure out where things fails. I have never done this on windows, but https://stackoverflow.com/questions/305287/how-to-see-the-contents-of-windows-library-lib looks like a good resource. The libraries you want to look at are are `C:\BLAS\lib\openblas.lib` and `amici.lib`, which should be located in `./build/` in the location where you are installing amici. Its easier to track this directory down when using an in-source installation as explained in https://amici.readthedocs.io/en/latest/python_installation.html#installation-of-development-versions.

**Comment by thomassligon** [comment]
On Windows with MSVC, there is a tool for checking dependencies. It is called dumpbin.exe. After running
`call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat"`
this tool is in my path:
```
C:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10>where dumpbin.exe
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\bin\Hostx64\x64\dumpbin.exe
```
Then you can find the options by running
`dumpbin -?`

Here is part of a previous discussion. I think this is covered by the documentation here:
[https://amici.readthedocs.io/en/latest/python_installation.html](https://amici.readthedocs.io/en/latest/python_installation.html)

The compilation and linking of openblas, including openblas.lib and openblas.dll, based purely on the openblas cmake files, is now working in Travis CI (#1152). After that, I installed this version of amici and tried it in my asthma PETab script (PE.py), which is currently just a beginning, i.e. a test. With that, I discovered that I need to have openblas.dll on my disk and in scope of the PATH variable in order for import amici to work.
This is because the dependencies of _amici.cp37-win_amd64.pyd are

    openblas.dll
    python37.dll
    MSVCP140.dll
    KERNEL32.dll
    VCRUNTIME140_1.dll
    VCRUNTIME140.dll
    api-ms-win-crt-heap-l1-1-0.dll
    api-ms-win-crt-stdio-l1-1-0.dll
    api-ms-win-crt-string-l1-1-0.dll
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-time-l1-1-0.dll
    api-ms-win-crt-math-l1-1-0.dll
and the dependencies of openblas.dll are

    KERNEL32.dll
    VCRUNTIME140.dll
    api-ms-win-crt-math-l1-1-0.dll
    api-ms-win-crt-stdio-l1-1-0.dll
    api-ms-win-crt-heap-l1-1-0.dll
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-convert-l1-1-0.dll
In other words, all dependencies lead to dlls which are supplied as part of Windows or Visual Studio, as long as openblas.dll is available.
That means that the user of Windows + VS (MSVC) would need to run the script installOpenBLAS.ps1, which calls compileBLAS.cmd. I think this takes us to the point where we need to make a final decision on whether this is the right way to do it.

**Comment by arnabmutsuddy** [comment]
Update: I was able to install amici after changing

`BLAS_LIBS="/LIBPATH:C:\BLAS\lib openblas.lib"`
to:
`BLAS_LIBS="/LIBPATH:C:\BLAS\lib" openblas.lib`

Now amici 0.11.16 shows up in the list of packages in pip. However, attempting to import amici in a python console causes the following error:
[error_log_spyder.txt](https://github.com/AMICI-dev/AMICI/files/6491021/error_log_spyder.txt)

I tried to follow the examples described in [this thread](https://github.com/AMICI-dev/AMICI/issues/1084), but it seems to involve a different version of amici and I couldn't figure out exactly what changes I'd need make in my version. 

In case it's still of interest, here's the result from running dumpbin on _amici.cp38-win_amd64.pyd
```

Dump of file _amici.cp38-win_amd64.pyd

File Type: DLL

  Image has the following dependencies:

    openblas.dll
    python38.dll
    MSVCP140.dll
    KERNEL32.dll
    VCRUNTIME140_1.dll
    VCRUNTIME140.dll
    api-ms-win-crt-heap-l1-1-0.dll
    api-ms-win-crt-stdio-l1-1-0.dll
    api-ms-win-crt-math-l1-1-0.dll
    api-ms-win-crt-string-l1-1-0.dll
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-time-l1-1-0.dll

  Summary

       1F000 .data
       18000 .pdata
       AB000 .rdata
        2000 .reloc
        1000 .rsrc
      1F1000 .text

```

**Comment by FFroehlich** [comment]
Okay, I am hopeful thats the last issue we have to fix. Thanks for seeing this through, it really helps us improving the documentations and prevent issues for future users.

My current suspicion is that some directory is missing from `PATH` and thats why the import fails at runtime. To fix this, I recommend doing the following steps in sequence:

1) double check whether `/LIBPATH:C:\BLAS\lib` was added to your `PATH` environment variable?

2) run dumpbin on `openblas.dll` and check that you get the following output:

```
    KERNEL32.dll
    VCRUNTIME140.dll
    api-ms-win-crt-math-l1-1-0.dll
    api-ms-win-crt-stdio-l1-1-0.dll
    api-ms-win-crt-heap-l1-1-0.dll
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-convert-l1-1-0.dll
```

3) check whether you find all of the dlls reported by dumpbin for either file in one of the directories listed in `PATH`

4) open `_amici.cp38-win_amd64.pyd` using https://www.dependencywalker.com and see whether that reports any kind of issue

**Comment by arnabmutsuddy** [comment]
> double check whether /LIBPATH:C:\lib was added to your PATH environment variable?

Are you sure this is what I should add to PATH? Because windows won't let me add anything with a forward slash (/) to PATH.

I already ran dumpbin on openblas.dll earlier and was able to find all the associated dlls. 
openblas.dll is located at C:\BLAS\bin for me and that has been added to PATH. But for some reason my python environment seems to be unable to find openblas.dll and I think that's causing the issue. However, I found a quick fix. I saw that when openblas.dll is copied from C:\BLAS\bin to amici installation directory (C:\Users\Arnab\AppData\Local\Programs\Python\Python38\Lib\site-packages\amici) amici can be imported to python and all of its functions can be used as usual.

**Comment by FFroehlich** [comment]
> > double check whether /LIBPATH:C:\lib was added to your PATH environment variable?
> 
> Are you sure this is what I should add to PATH? Because windows won't let me add anything with a forward slash (/) to PATH.

No, I am unfortunately not at all familiar with windows nomenclature or path handling.

> I already ran dumpbin on openblas.dll earlier and was able to find all the associated dlls.
> openblas.dll is located at C:\BLAS\bin for me and that has been added to PATH. But for some reason my python environment seems to be unable to find openblas.dll and I think that's causing the issue. However, I found a quick fix. I saw that when openblas.dll is copied from C:\BLAS\bin to amici installation directory (C:\Users\Arnab\AppData\Local\Programs\Python\Python38\Lib\site-packages\amici) amici can be imported to python and all of its functions can be used as usual.

Interesting. That definitely suggests that `openblas.dll` is not found on the path. Is openblas.dll contained in `C:\BLAS\lib` or `C:\BLAS\bin`, either way that directory needs to be added to path. You can check whether this is reflected in your python environment (I think it should), you can do something like

```
import os
print(os.os.environ['PATH'])
```

**Comment by thomassligon** [comment]
Yes, my `PATH` environment includes `C:\BLAS\bin`, which is where my openBLAS installation put `openplas.dll`:
```
C:\Users\thoma>where openblas.dll
C:\BLAS\bin\openblas.dll
```
Then, there are two special environment variables, on my machine they are
```
BLAS_CFLAGS=-IC:\\BLAS\\OpenBLAS-v0.3.10\\OpenBLAS-0.3.10
BLAS_LIBS=/LIBPATH:C:\\BLAS\\lib openblas.lib
```
These are used by the installation parts of AMICI to find `blas.h` to be included in `amici.cpp`, and `openblas.lib` to be linked into AMICI. I think they are working correctly on your machine. Then, a remark on command-line parameters. Traditionally, they start with a `-` on Unix and a `/` on Windows. However, today, most Windows commands accept either one. The same is true for directories, where, traditionally, Unix uses `/` and Windows uses `\`, and today most Windows commands accept either one. In the version of the environment variables that I am using, this is `\\`, because the variable is used as a string in python, so the directory separator needs to be written `\\` to result in `\`. I assume that `/` would work, but since I was searching for the cause to an error, I tried different variants, and this is just the one that I ended up with. Finally, `BLAS_CFLAGS` is used for compilation, the `-I` means include, and tells the compiler where to find `blas.h`. `BLAS_LIBS` is used for linking, `/LIBPATH`  means library path, and tells the linker where to find `openblas.lib`.

**Comment by thomassligon** [comment]
I see that I need to make a small correction. In your `amici_error_log2.txt`, I can see that `cl.exe` is being called to compile amici `cblas.cpp`, and it includes `/IC:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10`. So, this is where the environment variable is being used by `setup.py` to compile `amici\src\cblas.cpp`, an that needs an include for `cblas.h`. Whether you use `-I` or `/I` should be irrelevant, and whether you use `-LIBPATH` or `/LIBPATH` should also be irrelevant.

**Comment by arnabmutsuddy** [comment]
@thomassligon Thanks for the clarification. I agree with you that BLAS_CFLAGS and BLAS_LIBS are now working correctly in my machine, since the installation was successful. I also made sure the location of `openblas.dll` (`C:\BLAS\bin`) is included in my path and it also appeared when I looked up `os.environ['PATH']` from a python console. However, python was still unable to find openblas.dll unless it was located in the same folder as amici. I searched on the internet for a solution to this and came across some helpful information [here](https://docs.python.org/3/library/os.html#os.add_dll_directory). From what I understand, starting with python 3.8, changes have been made in the way how python looks for DLLs. So I ran `os.add_dll_directory("C:\\BLAS\\bin")` prior to `import amici` and that seems to solve the issue temporarily. But currently I'm not sure if there's a more permanent solution with the use of any other environment variables.

**Comment by dweindl** [comment]
> From what I understand, starting with python 3.8, changes have been made in the way how python looks for DLLs. 

Thanks for sharing. Good to know.

> But currently I'm not sure if there's a more permanent solution with the use of any other environment variables.

I guess that's something we could handle inside AMICI in the future. Not sure if it makes sense to just add all directories on %PATH% there (which might mess up the environment for other packages), or whether we should use a dedicated environment variable for AMICI.
